nav2_bt_navigator/src/bt_navigator.cpp

```diff
@@ -275,11 +280,31 @@ BtNavigator::navigateToPose()
       geometry_msgs::msg::PoseStamped goal_pose;
       blackboard_->get("goal", goal_pose);
 
-      feedback_msg->distance_remaining = nav2_util::geometry_utils::euclidean_distance(
-        feedback_msg->current_pose.pose, goal_pose.pose);
+      // Get current path points
+      nav_msgs::msg::Path current_path;
+      blackboard_->get("path", current_path);
+
+      // Calculate distance on the path
+      double distance_remaining =
+        nav2_util::geometry_utils::calculate_path_length(current_path);
```

Comment:
This returns the total path length, so how is this `distance_remaining`?

Not too familiar with the codebase yet, might be that `current_path` is a bit misleading name and actually means `path_left` :slightly_smiling_face:

Reply:
The path on the blackboard updates over time. So the `current_path` is essentially path remaining.

Reply:
But I could potentially find the point closest to the current pose of the robot and then calculate the distance based off of that pose, not sure if that's actually better though.

Reply:
> The path on the blackboard updates over time. So the current_path is essentially path remaining.

> But I could potentially find the point closest to the current pose of the robot and then calculate the distance based off of that pose

What is the difference between these two implementations?

Reply:
I suppose it depends - if the BT has absolutely no replanning, then yeah this would be never changing. However if there is replanning, then the `path` gets updated at the rate / time / distance in which the plan is being updated at.

We could do something like what we do in the regulated pure pursuit controller where we prune the path we already passed, but what do you think @Timple ? Is assuming replanning and updating this data at that rate sufficient?

Reply:
> Is assuming replanning and updating this data at that rate sufficient?

Not really, **exactly** in the case of using the pure pursuit controller. The main reason to use the pure pursuit controller would be to stick to a global plan as good as possible. So then I don't expect any replans...

Or am I messing up the terminology here?

Reply:
From what I observed the `path` on the blackboard got updated at roughly 2 Hz, which resulted in pretty accurate action feedback.

Reply:
> at roughly 2 Hz

That would be okay I guess. Are there any parameters that influence this 2hz? And can we have situations where it is actually 0hz?

Reply:
It all depends on the behavior tree used and what parameters you throw into it. All the BTs in nav2 by default have some type of consistent replanning based on time, distance traveled, or proportional rates to speed. It's certainly possible to have a situation where the path isn't updated, but none of the default trees do so. I think the default tree replans at 1hz.

I think it might be a good idea to prune the path in a local variable (read: don't update the blackboard path) and use that pruned path for this computation so you can get relatively frequent updated ETAs even with inconsistent replanning. 

What do you think @Timple ? Would it be reasonable to assume that the ETA/distances update at planning rate?

Reply:
You're referring to the planning rate of the local planner right?
In that case that would be fine yes.

Reply:
No, global planner

Reply:
But as I mentioned. The global planner simply might not have an update rate. Sometimes we just want to follow 'exactly this plan'. Without any replanning.
In that case the ETA/distances would never be updated.

Reply:
So then maybe pruning on this update is valuable. You already have the current pose, so with that and the path, you can find the nearest point to the robot, and throw that to the end of the path to the util function instead. You can see an example of a lambda that does this in robot base frame in the regulated pure pursuit controller.

Reply:
I second this pruning approach. I have a couple of use cases (bt setted) where the global plan is rarely if not never updated